### PR TITLE
Security Group Listing

### PR DESCRIPTION
**Pull request by Joeper214**
ref: https://skyarch.backlog.jp/view/NC-145

### Goal

- [x] Retrieve the list of security groups associated with the instance.
- [x] Display the list on UI
- [x] iT18N

### Next goal

* Change security groups associated with the instance
* Edit a security Group
* Add a new Security group
* Remove a security Group

### CODE REVIEW

File: app/controllers/nodes_controller.rb [hunk 1]
```diff
@@ -235,6 +237,17 @@ def edit_attributes
     @current_attributes = node.get_attributes
   end
 
+  # GET /nodes/:id/get_rules
+  def get_rules
+    physical_id = params.require(:id)
```
Comment (skyarch-networks) [inline app/controllers/nodes_controller.rb]: __[rubocop]__
* Lint/UselessAssignment: Useless assignment to variable - `physical_id`. [:ref](https://github.com/bbatsov/ruby-style-guide#underscore-unused-vars)

File: app/controllers/nodes_controller.rb [hunk 2]
```diff
@@ -235,6 +237,17 @@ def edit_attributes
...
+    group_ids = params.require(:group_ids)
+
+    rules_summary = @infra.ec2.describe_security_groups({group_ids: group_ids})
+
+    puts rules_summary.inspect
```
Comment (skyarch-networks) [inline app/controllers/nodes_controller.rb]: __[rubocop]__
* Rails/Output: Do not write to stdout. Use Rails' logger if you want to log.

File: app/controllers/nodes_controller.rb [hunk 3]
```diff
@@ -235,6 +237,39 @@ def edit_attributes
...
   end
 
+  # GET /nodes/:id/get_rules
+  def get_rules
+    group_ids = params.require(:group_ids)
+
+    rules_summary = @infra.ec2.describe_security_groups({group_ids: group_ids})
+
+    rules_summary[:security_groups].map do |item|
+       item
```
Comment (skyarch-networks) [inline app/controllers/nodes_controller.rb]: __[rubocop]__
* Style/IndentationWidth: Use 2 (not 3) spaces for indentation. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)
* Lint/Void: Variable `item` used in void context.

File: app/controllers/nodes_controller.rb [hunk 4]
```diff
@@ -235,6 +237,39 @@ def edit_attributes
...
+       item.ip_permissions.map do |inbound|
+         inbound
```
Comment (skyarch-networks) [inline app/controllers/nodes_controller.rb]: __[rubocop]__
* Lint/Void: Variable `inbound` used in void context.

File: app/controllers/nodes_controller.rb [hunk 5]
```diff
@@ -235,6 +237,39 @@ def edit_attributes
...
+         if inbound.from_port == -1 || inbound.from_port == nil || inbound.from_port == 0
+           inbound.user_id_group_pairs = 'All'
+         elsif inbound.from_port == 5439
+           inbound.user_id_group_pairs = 'Redshift'
+         else
+           inbound.user_id_group_pairs = Socket.getservbyport(inbound.from_port)
+         end
+       end
+       item.ip_permissions_egress.map do |outbound|
+         outbound
```
Comment (skyarch-networks) [inline app/controllers/nodes_controller.rb]: __[rubocop]__
* Lint/Void: Variable `outbound` used in void context.

File: app/controllers/nodes_controller.rb [hunk 6]
```diff
@@ -235,6 +237,36 @@ def edit_attributes
...
   end
 
+  # GET /nodes/:id/get_rules
+  def get_rules
+    group_ids = params.require(:group_ids)
+
+    rules_summary = @infra.ec2.describe_security_groups({group_ids: group_ids})
+
+    rules_summary[:security_groups].map do |item|
+       item.ip_permissions.map do |inbound|
```
Comment (skyarch-networks) [inline app/controllers/nodes_controller.rb]: __[rubocop]__
* Style/IndentationWidth: Use 2 (not 3) spaces for indentation. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)

File: app/controllers/nodes_controller.rb [hunk 7]
```diff
@@ -235,6 +237,36 @@ def edit_attributes
...
   end
 
+  # GET /nodes/:id/get_rules
+  def get_rules
+    group_ids = params.require(:group_ids)
+
+    rules_summary = @infra.ec2.describe_security_groups({group_ids: group_ids})
+
+    rules_summary[:security_groups].map do |item|
+     item.ip_permissions.map do |inbound|
```
Comment (skyarch-networks) [inline app/controllers/nodes_controller.rb]: __[rubocop]__
* Style/IndentationWidth: Use 2 (not 1) spaces for indentation. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)

**Comment by Joeper214** [comment]
Please review and Merge PR. @pocke  ..  Thanks!.. :bow:

**Comment by Joeper214** [comment]
Test is failed. I don't know the cause. Please help. @pocke .. :bow:

**Comment by pocke** [comment]
I think we cant build ruby-head. 
Our code doesn't have any problem.

**Comment by pocke** [comment]
I encounter an error when I click `view rule` button.

**Comment by Joeper214** [comment]
I also encountered this problem. but It seems so mysterious. because jQuery error message lacks details. but I'll find a solution. Thanks.

**Comment by pocke** [comment]
SkyHopper returns 500 error when port number is unknown.

**Comment by Joeper214** [comment]
I see. okay. I'll have a try catch on this. maybe ignore an unknown port? what do you think?

**Comment by pocke** [comment]
I think SkyHopper should show as "unknown" 

e.g.

**Comment by Joeper214** [comment]
:+1:  got it thanks.

**Comment by pocke** [comment]
Table is broken ... 


I think padding is not enough

**Comment by Joeper214** [comment]
Okay. I'll fix it. I think there's something missing on the table fillers. thanks

**Comment by pocke** [comment]
What's `-1` protocol ?

File: app/controllers/nodes_controller.rb [hunk 8]
```diff
@@ -235,6 +237,45 @@ def edit_attributes
...
+       end
+     end
+     item.ip_permissions_egress.map do |outbound|
+       if outbound.from_port == -1 || outbound.from_port == nil || outbound.from_port == 0
+         outbound.user_id_group_pairs = 'All'
+       elsif outbound.from_port == 5439
+         outbound.user_id_group_pairs = 'Redshift'
+       else
+         begin
+           outbound.user_id_group_pairs = Socket.getservbyport(outbound.from_port)
```
Comment (pocke) [inline app/controllers/nodes_controller.rb]: Please specify tcp or udp 
http://ruby-doc.org/stdlib-2.2.0/libdoc/socket/rdoc/Socket.html#method-c-getservbyport

File: app/controllers/nodes_controller.rb [hunk 9]
```diff
@@ -235,6 +237,45 @@ def edit_attributes
...
+         rescue
+           outbound.user_id_group_pairs = 'Unknown'
+         end
+       end
+     end
+    end
```
Comment (pocke) [inline app/controllers/nodes_controller.rb]: Can you make to DRY this method?

**Comment by pocke** [comment]
I apply the above security group.




Is this ok ?

File: app/controllers/nodes_controller.rb (hunk 8 above)
Comment (pocke) [inline app/controllers/nodes_controller.rb]: We should consider ICMP or custom protocol

File: app/controllers/nodes_controller.rb (hunk 9 above)
Comment (Joeper214) [inline app/controllers/nodes_controller.rb]: :+1:

File: app/controllers/nodes_controller.rb (hunk 8 above)
Comment (Joeper214) [inline app/controllers/nodes_controller.rb]: :+1:

**Comment by Joeper214** [comment]
The table is okay. but there are some fields that are mismatched.

File: app/controllers/nodes_controller.rb [hunk 10]
```diff
@@ -235,6 +237,38 @@ def edit_attributes
...
   end
 
+  # GET /nodes/:id/get_rules
+  def get_rules
+    group_ids = params.require(:group_ids)
+
+    rules_summary = @infra.ec2.describe_security_groups({group_ids: group_ids})
+
+    rules_summary[:security_groups].map do |item|
+     check_socket(item.ip_permissions)
```
Comment (skyarch-networks) [inline app/controllers/nodes_controller.rb]: __[rubocop]__
* Style/IndentationWidth: Use 2 (not 1) spaces for indentation. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)

File: app/controllers/nodes_controller.rb [hunk 11]
```diff
@@ -235,6 +237,37 @@ def edit_attributes
     @current_attributes = node.get_attributes
   end
 
+  # GET /nodes/:id/get_rules
+  def get_rules
+    group_ids = params.require(:group_ids)
+    rules_summary = @infra.ec2.describe_security_groups({group_ids: group_ids})
+
+    rules_summary[:security_groups].map do |item|
+     check_socket(item.ip_permissions)
```
Comment (skyarch-networks) [inline app/controllers/nodes_controller.rb]: __[rubocop]__
* Style/IndentationWidth: Use 2 (not 1) spaces for indentation. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)

**Comment by Joeper214** [comment]
`What is -1 protocol` https://github.com/skyarch-networks/skyhopper/pull/122#issuecomment-159798493

 - It should be -1 = 'All'    I think your branch is not updated.

**Comment by Joeper214** [comment]
Just like this.





wed

**Comment by pocke** [comment]
:+1:

**Comment by Joeper214** [comment]
Thanks.

**Pull request merged**